File: lib/biz/spec-iot/src/main/java/site/addzero/biz/spec/iot/protocol/modbus/ModbusClient.java
package site.addzero.biz.spec.iot.protocol.modbus;

import com.serotonin.modbus4j.BatchRead;
import com.serotonin.modbus4j.BatchResults;
import com.serotonin.modbus4j.ModbusFactory;
import com.serotonin.modbus4j.ModbusMaster;
import com.serotonin.modbus4j.code.DataType;
import com.serotonin.modbus4j.ip.IpParameters;
import com.serotonin.modbus4j.locator.BaseLocator;
import com.serotonin.modbus4j.msg.WriteCoilRequest;
import com.serotonin.modbus4j.msg.WriteCoilResponse;
import com.serotonin.modbus4j.msg.WriteRegisterRequest;
import com.serotonin.modbus4j.msg.WriteRegisterResponse;

import java.util.LinkedHashMap;
import java.util.List;
import java.util.Map;
import java.util.concurrent.ConcurrentHashMap;

/**
 * Thin Modbus TCP client wrapper with connection caching.
 */
public final class ModbusClient {

    private static final ModbusFactory FACTORY = new ModbusFactory();
    private static final Map<String, ModbusMaster> MASTERS = new ConcurrentHashMap<String, ModbusMaster>();

    public ModbusMaster getMaster(ModbusConnectionOptions options) {
        ModbusMaster existing = MASTERS.get(options.cacheKey());
        if (existing != null) {
            return existing;
        }
        synchronized (ModbusClient.class) {
            ModbusMaster current = MASTERS.get(options.cacheKey());
            if (current != null) {
                return current;
            }
            try {
                IpParameters parameters = new IpParameters();
                parameters.setHost(options.getHost());
                parameters.setPort(options.getPort());
                ModbusMaster created = FACTORY.createTcpMaster(parameters, false);
                created.init();
                MASTERS.put(options.cacheKey(), created);
                return created;
            } catch (Exception ex) {
                throw new IllegalStateException("Failed to initialize Modbus master", ex);
            }
        }
    }

    public void close(ModbusConnectionOptions options) {
        ModbusMaster existing = MASTERS.remove(options.cacheKey());
        if (existing != null) {
            try {
                existing.destroy();
            } catch (Exception ignored) {
                // Ignore close failures.
            }
        }
    }

    public boolean verifyConnection(ModbusConnectionOptions options) {
        try {
            BatchRead<String> testBatch = new BatchRead<String>();
            testBatch.addLocator("health_check", BaseLocator.inputRegister(1, 0, DataType.TWO_BYTE_INT_UNSIGNED));
            testBatch.setContiguousRequests(true);
            getMaster(options).send(testBatch);
            return true;
        } catch (Exception ignored) {
            return false;
        }
    }

    public Map<String, Object> batchReadRaw(ModbusConnectionOptions options, List<ModbusPointBinding> points) {
        BatchRead<String> batch = new BatchRead<String>();
        for (ModbusPointBinding point : points) {
            if (point.getRegisterType() == ModbusRegisterType.COIL_STATUS) {
                batch.addLocator(point.getPointId(), BaseLocator.coilStatus(point.getSlaveId(), point.getPointAddress()));
            } else if (point.getRegisterType() == ModbusRegisterType.INPUT_STATUS) {
                batch.addLocator(point.getPointId(), BaseLocator.inputStatus(point.getSlaveId(), point.getPointAddress()));
            } else if (point.getRegisterType() == ModbusRegisterType.HOLDING_REGISTER) {
                batch.addLocator(point.getPointId(), BaseLocator.holdingRegister(point.getSlaveId(), point.getPointAddress(), DataType.FOUR_BYTE_FLOAT_SWAPPED));
            } else if (point.getRegisterType() == ModbusRegisterType.INPUT_REGISTER) {
                batch.addLocator(point.getPointId(), BaseLocator.inputRegister(point.getSlaveId(), point.getPointAddress(), DataType.FOUR_BYTE_FLOAT_SWAPPED));
            }
        }

        try {
            batch.setContiguousRequests(true);
            BatchResults<String> results = getMaster(options).send(batch);
            Map<String, Object> rawValues = new LinkedHashMap<String, Object>();
            for (ModbusPointBinding point : points) {
                rawValues.put(point.getPointId(), results.getValue(point.getPointId()));
            }
            return rawValues;
        } catch (Exception ex) {
            throw new IllegalStateException("Failed to batch read Modbus points", ex);
        }
    }

    public boolean singleWritePoint(ModbusConnectionOptions options, ModbusWritePoint point) {
        try {
            ModbusMaster master = getMaster(options);
            if (point.getRegisterType() == ModbusRegisterType.COIL_STATUS) {
                WriteCoilResponse response = (WriteCoilResponse) master.send(
                        new WriteCoilRequest(point.getSlaveId(), point.getPointAddress(), Boolean.parseBoolean(point.getWriteValue()))
                );
                return !response.isException();
            }
            if (point.getRegisterType() == ModbusRegisterType.HOLDING_REGISTER) {
                String writeValue = point.getWriteValue();
                if (point.getValueType() == site.addzero.biz.spec.iot.IotValueType.BOOLEAN) {
                    writeValue = Boolean.parseBoolean(writeValue) ? "1" : "0";
                }
                WriteRegisterResponse response = (WriteRegisterResponse) master.send(
                        new WriteRegisterRequest(point.getSlaveId(), point.getPointAddress(), Integer.parseInt(writeValue))
                );
                return !response.isException();
            }
            throw new IllegalArgumentException("Unsupported Modbus write register type: " + point.getRegisterType());
        } catch (Exception ex) {
            throw new IllegalStateException("Failed to write Modbus point " + point.getPointId(), ex);
        }
    }
}

File: lib/biz/spec-iot/src/main/kotlin/site/addzero/biz/spec/iot/protocol/modbus/ModbusClient.kt
package site.addzero.biz.spec.iot.protocol.modbus

import site.addzero.biz.spec.iot.IotValueType
import java.lang.reflect.Field
import java.util.concurrent.ConcurrentHashMap

/**
 * Thin Modbus TCP client wrapper with connection caching.
 */
class ModbusClient {

    fun close(options: ModbusConnectionOptions) {
        val existing = MASTERS.remove(options.cacheKey())
        if (existing != null) {
            try {
                existing.javaClass.getMethod("destroy").invoke(existing)
            } catch (_: Exception) {
                // Ignore close failures.
            }
        }
    }

    fun verifyConnection(options: ModbusConnectionOptions): Boolean {
        return try {
            val batchReadClass = loadClass(BATCH_READ_CLASS_NAME)
            val batch = batchReadClass.getConstructor().newInstance()
            val locator = buildInputRegisterLocator(1, 0, "TWO_BYTE_INT_UNSIGNED")
            batchReadClass.getMethod("addLocator", Any::class.java, Any::class.java).invoke(batch, "health_check", locator)
            batchReadClass.getMethod("setContiguousRequests", Boolean::class.javaPrimitiveType).invoke(batch, true)
            getMaster(options).javaClass.getMethod("send", batchReadClass).invoke(getMaster(options), batch)
            true
        } catch (_: Exception) {
            false
        }
    }

    fun batchReadRaw(options: ModbusConnectionOptions, points: List<ModbusPointBinding>): Map<String, Any?> {
        try {
            val batchReadClass = loadClass(BATCH_READ_CLASS_NAME)
            val batch = batchReadClass.getConstructor().newInstance()
            points.forEach { point ->
                val locator = buildLocator(point)
                batchReadClass.getMethod("addLocator", Any::class.java, Any::class.java)
                    .invoke(batch, point.pointId, locator)
            }
            batchReadClass.getMethod("setContiguousRequests", Boolean::class.javaPrimitiveType).invoke(batch, true)
            val results = getMaster(options).javaClass.getMethod("send", batchReadClass).invoke(getMaster(options), batch)
            val getValue = loadClass(BATCH_RESULTS_CLASS_NAME).getMethod("getValue", Any::class.java)
            val rawValues = linkedMapOf<String, Any?>()
            points.forEach { point ->
                rawValues[point.pointId] = getValue.invoke(results, point.pointId)
            }
            return rawValues.toMap()
        } catch (ex: Exception) {
            throw IllegalStateException("Failed to batch read Modbus points", ex)
        }
    }

    fun singleWritePoint(options: ModbusConnectionOptions, point: ModbusWritePoint): Boolean {
        try {
            val master = getMaster(options)
            if (point.registerType == ModbusRegisterType.COIL_STATUS) {
                val request = loadClass(WRITE_COIL_REQUEST_CLASS_NAME)
                    .getConstructor(Int::class.javaPrimitiveType, Int::class.javaPrimitiveType, Boolean::class.javaPrimitiveType)
                    .newInstance(point.slaveId, point.pointAddress, java.lang.Boolean.parseBoolean(point.writeValue))
                val response = master.javaClass
                    .getMethod("send", loadClass("com.serotonin.modbus4j.msg.ModbusRequest"))
                    .invoke(master, request)
                return !(response.javaClass.getMethod("isException").invoke(response) as Boolean)
            }
            if (point.registerType == ModbusRegisterType.HOLDING_REGISTER) {
                var writeValue = requireWriteValue(point.writeValue)
                if (point.valueType == IotValueType.BOOLEAN) {
                    writeValue = if (java.lang.Boolean.parseBoolean(writeValue)) {
                        "1"
                    } else {
                        "0"
                    }
                }
                val request = loadClass(WRITE_REGISTER_REQUEST_CLASS_NAME)
                    .getConstructor(Int::class.javaPrimitiveType, Int::class.javaPrimitiveType, Int::class.javaPrimitiveType)
                    .newInstance(point.slaveId, point.pointAddress, Integer.parseInt(writeValue))
                val response = master.javaClass
                    .getMethod("send", loadClass("com.serotonin.modbus4j.msg.ModbusRequest"))
                    .invoke(master, request)
                return !(response.javaClass.getMethod("isException").invoke(response) as Boolean)
            }
            throw IllegalArgumentException("Unsupported Modbus write register type: ${point.registerType}")
        } catch (ex: Exception) {
            throw IllegalStateException("Failed to write Modbus point ${point.pointId}", ex)
        }
    }

    private fun getMaster(options: ModbusConnectionOptions): Any {
        val existing = MASTERS[options.cacheKey()]
        if (existing != null) {
            return existing
        }
        return synchronized(MASTERS) {
            MASTERS[options.cacheKey()] ?: run {
                try {
                    createMaster(options).also { MASTERS[options.cacheKey()] = it }
                } catch (ex: Exception) {
                    throw IllegalStateException("Failed to initialize Modbus master", ex)
                }
            }
        }
    }

    private fun createMaster(options: ModbusConnectionOptions): Any {
        val factoryClass = loadClass(FACTORY_CLASS_NAME)
        val factory = factoryClass.getConstructor().newInstance()
        val ipParametersClass = loadClass(IP_PARAMETERS_CLASS_NAME)
        val parameters = ipParametersClass.getConstructor().newInstance()
        ipParametersClass.getMethod("setHost", String::class.java).invoke(parameters, options.host)
        ipParametersClass.getMethod("setPort", Int::class.javaPrimitiveType).invoke(parameters, options.port)
        val master = factoryClass.getMethod("createTcpMaster", ipParametersClass, Boolean::class.javaPrimitiveType)
            .invoke(factory, parameters, false)
        loadClass(MASTER_CLASS_NAME).getMethod("init").invoke(master)
        return master
    }

    private fun buildLocator(point: ModbusPointBinding): Any {
        if (point.registerType == ModbusRegisterType.COIL_STATUS) {
            return loadClass(BASE_LOCATOR_CLASS_NAME)
                .getMethod("coilStatus", Int::class.javaPrimitiveType, Int::class.javaPrimitiveType)
                .invoke(null, point.slaveId, point.pointAddress)
        }
        if (point.registerType == ModbusRegisterType.INPUT_STATUS) {
            return loadClass(BASE_LOCATOR_CLASS_NAME)
                .getMethod("inputStatus", Int::class.javaPrimitiveType, Int::class.javaPrimitiveType)
                .invoke(null, point.slaveId, point.pointAddress)
        }
        if (point.registerType == ModbusRegisterType.HOLDING_REGISTER) {
            return loadClass(BASE_LOCATOR_CLASS_NAME)
                .getMethod("holdingRegister", Int::class.javaPrimitiveType, Int::class.javaPrimitiveType, loadClass(DATA_TYPE_CLASS_NAME))
                .invoke(null, point.slaveId, point.pointAddress, dataTypeConstant("FOUR_BYTE_FLOAT_SWAPPED"))
        }
        if (point.registerType == ModbusRegisterType.INPUT_REGISTER) {
            return loadClass(BASE_LOCATOR_CLASS_NAME)
                .getMethod("inputRegister", Int::class.javaPrimitiveType, Int::class.javaPrimitiveType, loadClass(DATA_TYPE_CLASS_NAME))
                .invoke(null, point.slaveId, point.pointAddress, dataTypeConstant("FOUR_BYTE_FLOAT_SWAPPED"))
        }
        throw IllegalArgumentException("Unsupported Modbus register type: ${point.registerType}")
    }

    private fun buildInputRegisterLocator(slaveId: Int, pointAddress: Int, dataTypeField: String): Any {
        return loadClass(BASE_LOCATOR_CLASS_NAME)
            .getMethod("inputRegister", Int::class.javaPrimitiveType, Int::class.javaPrimitiveType, loadClass(DATA_TYPE_CLASS_NAME))
            .invoke(null, slaveId, pointAddress, dataTypeConstant(dataTypeField))
    }

    private fun dataTypeConstant(fieldName: String): Any? {
        val field: Field = loadClass(DATA_TYPE_CLASS_NAME).getField(fieldName)
        return field.get(null)
    }

    private fun loadClass(className: String): Class<*> {
        try {
            return Class.forName(className)
        } catch (ex: ClassNotFoundException) {
            throw IllegalStateException("Missing runtime dependency for Modbus support: $className", ex)
        }
    }

    private fun requireWriteValue(value: String?): String {
        return value ?: throw IllegalArgumentException("writeValue must not be null")
    }

    companion object {
        private const val FACTORY_CLASS_NAME = "com.serotonin.modbus4j.ModbusFactory"
        private const val MASTER_CLASS_NAME = "com.serotonin.modbus4j.ModbusMaster"
        private const val IP_PARAMETERS_CLASS_NAME = "com.serotonin.modbus4j.ip.IpParameters"
        private const val BATCH_READ_CLASS_NAME = "com.serotonin.modbus4j.BatchRead"
        private const val BATCH_RESULTS_CLASS_NAME = "com.serotonin.modbus4j.BatchResults"
        private const val BASE_LOCATOR_CLASS_NAME = "com.serotonin.modbus4j.locator.BaseLocator"
        private const val DATA_TYPE_CLASS_NAME = "com.serotonin.modbus4j.code.DataType"
        private const val WRITE_COIL_REQUEST_CLASS_NAME = "com.serotonin.modbus4j.msg.WriteCoilRequest"
        private const val WRITE_REGISTER_REQUEST_CLASS_NAME = "com.serotonin.modbus4j.msg.WriteRegisterRequest"
        private val MASTERS = ConcurrentHashMap<String, Any>()
    }
}
